docs/rules/destructuring-props-argument.md

```diff
@@ -0,0 +1,45 @@
+# Enforces consistent usage of destructuring props argument assignment(react/destructuring-props-argument)
+
+Rule enforces consistent usage of destructuring assignment in component arguments.
+
+Rule can be set to either of `always`, `never`, `ignore` for each type of the component.
+Currently only SFC is implemented.
+
+```js
+"react/destructuring-props-argument": [<enabled>, { "SFC": "always"}]
+```
+
+## Rule Details
+
+By default rule is set to `always` enforce destructuring assignment. The following pattern is considered warning:
```

Comment:
presumably we'd want the default to be "always" on all component types; releasing the rule with only SFC support means we couldn't default class and createClass components to "always" without a major release

Reply:
I am not sure about your major release concern. Would releasing only SFC support require a major release? What's the problem eith having to do a major release?

Irrespective of that I have only one use case for "other" component types as per issue #278 and I am not particularly keen about it.  Are you? Maybe it would make sense to have this rule for SFC only?

Reply:
releasing only SFC support would be a minor.

Later releasing class-based component supported, defaulted to "ignore", would also be a minor.

However, later releasing class-based component supported, defaulted to "always", would be a major.

There's not a *problem* per se about releasing a major release, but it's always best to minimize breaking changes.

My use case for non-SFC components is that I want *all* `this.props`, `this.state`, and `this.context` properties destructured at the top of every function that uses them (that also includes `context` in SFCs).

It would be great if one rule could handle all of these cases - if not, I'd expect this rule to only *ever* handle SFCs (props and context), and a new rule to be added for "inside non-SFC functions".

Reply:
Ok, well in this case I can  try to add the rest of the logic to this pr.
So both of the component types will enforce `context` and `props`. Class component will in addition enforce `state`. 
Config structure would be: `{ SFC: always/never/ignore,  class: always/never/ignore}` (I don't like name `class` is there a better term?)
default: `{ SFC: always, class: always }`

there is also a nextProps, prevProps which I would ignore for now.
Does this make sense?

Reply:
i think `class` is probably the best name, unfortunately.

I agree ignoring next/prev props is fine for now, in SFCs, it'd be `props` and `context`, and in class-based and createClass components, it'd be `this.props` and `this.state` and `this.context`.